File: c/graph.c
#include "graph.h"
#include <stdlib.h>
#include <stdio.h>

struct Graph * new_graph(int num_verticles) {
    struct Graph * graph = (struct Graph*)malloc(sizeof (struct Graph));
    graph->vertex_array = (struct Vertex*)calloc(num_verticles, sizeof (struct Vertex));
    graph->num_verticles = num_verticles;
    return graph;
}

void graph_free(struct Graph * g) {
    for (int i =0; i < g->num_verticles; i++) {
#ifdef DEBUG
        printf("V[%d]:\n", i);
#endif
        struct Vertex curr = g->vertex_array[i];
        struct Neighbor * neighbor = curr.neighbors;
        while (neighbor != NULL) {
#ifdef DEBUG
            printf("\t[E] %d --%d--> %d\n", i, neighbor->weight, neighbor->vertex_index);
#endif
            struct Neighbor * to_free = neighbor;
            neighbor = neighbor->next;
            free(to_free);
        }
    }
    free(g->vertex_array);
    free(g);
}

void graph_add_single_edge(struct Graph * g, int from, int to, int weight) {
    struct Vertex *f = &(g->vertex_array[from-1]);
    struct Neighbor *nforward = (struct Neighbor *)malloc(sizeof(struct Neighbor));
    nforward->vertex_index = to-1;
    nforward->weight = weight;
    nforward->next = NULL;
    if (f->neighbors == NULL) {
        f->neighbors = nforward;
    } else {
        nforward->next = f->neighbors;
        f->neighbors = nforward;
    }
}

void graph_add_edge(struct Graph * g, int from, int to, int weight) {
#ifdef DEBUG
    printf("graph_add_edge(%d, %d, %d)\n", from, to, weight);
#endif
    graph_add_single_edge(g, from, to, weight);
    graph_add_single_edge(g, to, from, weight);

}


struct Neighbor * graph_get_neighbor(struct Graph * g, unsigned int from) {
    struct Vertex f = g->vertex_array[from];
    return f.neighbors;
}

File: java/Graph.java
import java.util.LinkedList;

public class Graph
{
    private Vertex[] vertex_array;
    private int num_vertices;

    public int getNum_vertices()
    {
        return num_vertices;
    }

    public Graph(int n_ver)
    {
        vertex_array = new Vertex[n_ver];
        num_vertices = n_ver;

        for (int i = 0; i < n_ver; ++i)
        {
            vertex_array[i] = new Vertex();
        }
    }

    public void graph_add_single_edge(int from, int to, long weight)
    {
        Vertex f = vertex_array[from - 1];
        Neighbor nforward = new Neighbor();
        nforward.setVertex_index(to - 1);
        nforward.setWeight(weight);
        nforward.next = null;
        if (f.neighbors == null) {
            f.neighbors = nforward;
        } else {
            nforward.next = f.neighbors;
            f.neighbors = nforward;
        }
    }

    void graph_add_edge(int from, int to, long weight)
    {
        graph_add_single_edge(from, to, weight);
        graph_add_single_edge(to, from, weight);
    }

    Neighbor graph_get_neighbors(int from)
    {
        Vertex f = vertex_array[from];
        return f.neighbors;
    }
}

class Neighbor
{
    private int vertex_index;
    private long weight;

    public int getVertex_index()
    {
        return vertex_index;
    }

    public void setVertex_index(int vertex_index)
    {
        this.vertex_index = vertex_index;
    }

    public long getWeight()
    {
        return weight;
    }

    public void setWeight(long weight)
    {
        this.weight = weight;
    }
    public Neighbor next;
}

class Vertex
{
    public Neighbor neighbors;
}
